### PR TITLE
feat: generate templates from local/remote files

### PR DESCRIPTION
**Pull request by franky47**
Adds a generator behaviour for websites (template not yet implemented) and webapps (currently only Next.js blank, no Juno setup yet).

When developping locally, use the `CREATE_JUNO_USE_LOCAL_FILES=true` environment variable to use local template files from the filesystem (that may otherwise not yet exist on GitHub). Otherwise by default files are pulled from the GitHub CDN from this repository.

Also includes a detection of whether the Juno CLI is installed, which runs in parallel of the user flow to avoid delays when prompting after generation.

Left to do before merging:
- [ ] Include Juno setup in basic `nextjs-blank` example (straight out of `create next-app`)
- [ ] Add a `nextjs-tutorial` template with the diary example
- [ ] Add Astro website template and enable the website branch of the user flow
- [ ] Add more app templates using the same system

### CODE REVIEW

File: src/utils/populate.utils.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  );
+  return files
+    .flat()
+    .map((p) =>
+      typeof p === 'string' ? {path: relative(process.cwd(), p), url: `file://${p}`} : p
+    );
+};
+
+const getGitHubFiles = async (templatePath: string): Promise<FileDescriptor[]> => {
+  const res = await fetch(`https://api.github.com/repos/${REPOSITORY}/git/trees/main?recursive=1`);
```
Comment (peterpeterparker) [inline src/utils/populate.utils.ts:33]: We might easily hit some rate limits with the GitHub API given my experience.

Not for this PR, just thinking about it now, should we compress (?, gzip) and deploy the template (CI action) to a "CDN" (assets smart contract/website) and download those from there?

Alternatively, should we include the templates with the CLI?

I guess the latter has the tradeoff of potentially making the CLI quite heavy.

Again, not for this PR, which is already substantial, but for another iteration.

File: src/utils/populate.utils.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  return tree
+    .filter(({type, path: p}) => type === 'blob' && p.startsWith(templatePath))
+    .map(({path}) => ({
+      path,
+      url: `https://raw.githubusercontent.com/${REPOSITORY}/main/${path}`
+    }));
+};
+
+const createDirectory = async (where: string | null) => {
+  if (where === null) {
```
Comment (peterpeterparker) [inline src/utils/populate.utils.ts:52]: where can be null? should the user specify a destination or null === same directory?

File: src/utils/populate.utils.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+};
+
+interface PopulateInput {
+  where: string | null;
+  templatePath: string;
+}
+
+export const populate = async ({where, templatePath}: PopulateInput) => {
+  const spinner = ora(`Creating example...`).start();
+  const useLocalFiles = process.env.CREATE_JUNO_USE_LOCAL_FILES === 'true';
```
Comment (peterpeterparker) [inline src/utils/populate.utils.ts:74]: remind me to set up a `.env` for this project as I did in the main CLI.

File: src/utils/populate.utils.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+  try {
+    const files = await (useLocalFiles
+      ? getLocalFiles(templatePath)
+      : getGitHubFiles(templatePath));
+    console.dir({
```
Comment (peterpeterparker) [inline src/utils/populate.utils.ts:80]: those console output are leftovers?

File: src/commands/install.ts [hunk 5]
```diff
@@ -16,3 +22,29 @@ export const installCli = async () => {
 
   console.log(`\n✅ CLI installed. Run ${cyan('juno --help')} to display more information.`);
 };
+
+export async function installCliIfNecessary() {
```
Comment (peterpeterparker) [inline src/commands/install.ts:26]: for consistency reason, can you use an arrow function or any particular reason to use a function here?

File: src/commands/install.ts [hunk 6]
```diff
@@ -16,3 +22,29 @@ export const installCli = async () => {
...
+  });
+  if (!performInstall) {
+    return;
+  }
+  await installCli();
+}
+
+async function detectIfCliIsInstalled() {
+  const exitCode = await spawn({
+    command: 'npm',
```
Comment (peterpeterparker) [inline src/commands/install.ts:45]: Cool!

**Review by peterpeterparker — COMMENTED** [review]
Looks good 👍

Few minor questions

File: src/commands/generate.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  name: string;
+  template: Template;
+  starter: 'blank' | 'tutorial' | null;
+}
+
+export const generate = async ({action, name, starter, template}: GeneratorInput) => {
+  const templatePath = path.join(
+    'templates',
+    action,
+    template.key + (starter !== null ? `-${starter}` : '')
```
[the remark this answers is not in the document]
Comment (franky47) [reply src/commands/generate.ts:68]: I was wondering about the structure and naming, `framework-blank` is not particularly neat. But yeah I agree, whatever the convention, we can abstract that into a function.

File: src/utils/populate.utils.ts (hunk 2 above)
Comment (franky47) [reply src/utils/populate.utils.ts:52]: Yes, null implies the user chose the current directory as the destination, either by pressing Enter at the project name prompt (no input), or typing '.' (current directory).

There are two possible cases:
1. The user alread has created a destination directory, is inside it, and runs the CLI to populate (what I usually do)
2. The user lets the CLI create a directory for them, by naming it.

The "failure" cases here would be:
1. The CLI creates another subdirectory with the same name (resulting in `path/to/foo/foo/templateFilesHere`)
2. The CLI populates the CWD which is non-empty

Suggesting the CWD as a placeholder project name solves 1., to solve 2. we'd need to detect if the target directory exists and isn't empty.

What do you think?

File: src/utils/populate.utils.ts (hunk 1 above)
Comment (peterpeterparker) [reply src/utils/populate.utils.ts:33]: Let's use a CDN - i.e. Juno's CDN. I'll setup this.

File: src/utils/populate.utils.ts (hunk 2 above)
Comment (peterpeterparker) [reply src/utils/populate.utils.ts:52]: Thanks for the explanation. Let's start with the existing and iterate with some tests if needed.

**Comment by peterpeterparker** [comment]
Thanks!

**Comment by peterpeterparker** [comment]
I'll now setup env and cdn.

**Pull request merged**